File: src/DRandom.cs
/*******************************************************************************
*  Copyright by the contributors to the Dafny Project
*  SPDX-License-Identifier: MIT
*******************************************************************************/

using System;
using System.Numerics;

namespace DafnyLibraries {
  public partial class DRandomFoundational {
    private Random r;

    public DRandomFoundational() {
      this.r = new Random();
    }

    public bool Coin() {
      return this.r.Next(2) == 0;
    }
  }

  public partial class DRandomExternUniform {
    private Random r;

    public DRandomExternUniform() {
      this.r = new Random();
    }

    public bool Coin() {
      return this.r.Next(2) == 0;
    }

    public BigInteger Uniform(BigInteger n) {
      return new BigInteger(this.r.Next((int) n));
    }
  }  
}

File: src/DRandom.java
/*******************************************************************************
*  Copyright by the contributors to the Dafny Project
*  SPDX-License-Identifier: MIT
*******************************************************************************/

package DafnyLibraries;

import java.security.SecureRandom;
import java.math.BigInteger;

class DRandomFoundational extends _ExternBase_DRandomFoundational {
  private SecureRandom s;

  public DRandomFoundational() {
    this.s = new SecureRandom();
  }

  @Override
  public boolean Coin() {
    return s.nextBoolean();
  }
}

class DRandomExternUniform extends _ExternBase_DRandomExternUniform {
  private SecureRandom s;

  public DRandomExternUniform() {
    this.s = new SecureRandom();
  }

  @Override
  public boolean Coin() {
    return s.nextBoolean();
  }

  @Override
  public int Uniform(int n) {
    return s.nextInt(n);
  }
}

